# Lab1: part 3 - Data Types and Arrays

So far, we've used Python to manipulate numbers and work with dataframes.  We need to discuss data types to deepen our understanding of how to work with data in Python.

In this section, you'll first see how to represent and manipulate another fundamental type of data: text.  A piece of text is called a *string* in Python. You'll also see how to work with *arrays* of data, such as all the numbers between 0 and 100 or all the words in the chapter of a book. Lastly, you'll create df's and practice analyzing them with your knowledge of df operations.
 

Set up the tests and imports by running the cell below.

In [14]:
# Just run this cell
import numpy as np
import pandas as pd
import math


# 1. Text
Programming doesn't just concern numbers. Text is one of the most common data types used in programs. 

Text is represented by a **string value** in Python. The word "string" is a programming term for a sequence of characters. A string might contain a single character, a word, a sentence, or a whole book.

To distinguish text data from actual code, we demarcate strings by putting quotation marks around them. Single quotes (`'`) and double quotes (`"`) are both valid, but the types of opening and closing quotation marks must match. The contents can be any sequence of characters, including numbers and symbols. 

We've seen strings before in `print` statements.  Below, two different strings are passed as arguments to the `print` function.

In [15]:
print("I <3", 'Data Science')

I <3 Data Science


Just as names can be given to numbers, names can be given to string values.  The names and strings aren't required to be similar in any way. Any name can be assigned to any string.

In [16]:
one = 'two'
plus = '*'
print(one, plus, one)

two * two


**Question 1.1.** Yuri Gagarin was the first person to travel through outer space.  When he emerged from his capsule upon landing on Earth, he [reportedly](https://en.wikiquote.org/wiki/Yuri_Gagarin) had the following conversation with a woman and girl who saw the landing:

    The woman asked: "Can it be that you have come from outer space?"
    Gagarin replied: "As a matter of fact, I have!"

The cell below contains unfinished code.  Fill in the `...`s so that it prints out this conversation *exactly* as it appears above.

<!--
BEGIN QUESTION
name: q11
-->

In [17]:
woman_asking = "Can it be that you have come from outer space?"
woman_quote = '"Can it be that you have come from outer space?"'
gagarin_reply = 'Gagarin replied:'
gagarin_quote = "As a matter of fact, I have!"

print(woman_asking, woman_quote)
print(gagarin_reply, gagarin_quote)

Can it be that you have come from outer space? "Can it be that you have come from outer space?"
Gagarin replied: As a matter of fact, I have!


## 1.1. String Methods

Strings can be transformed using **methods**. Recall that methods and functions are not technically the same thing, but we'll be using them interchangeably for the purposes of this course.

Here's a sketch of how to call methods on a string:

    <expression that evaluates to a string>.<method name>(<argument>, <argument>, ...)
    
One example of a string method is `replace`, which replaces all instances of some part of the original string (or a *substring*) with a new string. 

    <original string>.replace(<old substring>, <new substring>)
    
`replace` returns (evaluates to) a new string, leaving the original string unchanged.
    
Try to predict the output of this example, then run the cell!

In [18]:
# Replace one letter
hello = 'Hello'
print(hello.replace('o', 'a'), hello)

Hella Hello


You can call functions on the results of other functions.  For example, `max(abs(-5), abs(3))` evaluates to 5.  Similarly, you can call methods on the results of other method or function calls.

You may have already noticed one difference between functions and methods - a function like `max` does not require a `.` before it's called, but a string method like `replace` does. Here's a handy [Python reference](http://data8.org/sp20/python-reference.html) on the Data 8 website. It's a good idea to refer to this whenever you're unsure of how to call a function or method.

In [19]:
# Calling replace on the output of another call to replace
'train'.replace('t', 'ing').replace('in', 'de')

'degrade'

Here's a picture of how Python evaluates a "chained" method call like that:

<img src="chaining_method_calls.png"/>

**Question 1.1.1.** Use `replace` to transform the string `'hitchhiker'` into `'matchmaker'`. Assign your result to `new_word`.

<!--
BEGIN QUESTION
name: q111
-->

In [20]:
new_word = 'hitchhiker'.replace('hitchhi','matchma')
new_word

'matchmaker'

There are many more string methods in Python, but most programmers don't memorize their names or how to use them.  In the "real world," people usually just search the internet for documentation and examples. A complete [list of string methods](https://docs.python.org/3/library/stdtypes.html#string-methods) appears in the Python language documentation. [Stack Overflow](http://stackoverflow.com) has a huge database of answered questions that often demonstrate how to use these methods to achieve various ends.

## 1.2. Converting to and from Strings

Strings and numbers are different *types* of values, even when a string contains the digits of a number. For example, evaluating the following cell causes an error because an integer cannot be added to a string.

In [21]:
8 + "8"

TypeError: unsupported operand type(s) for +: 'int' and 'str'

However, there are built-in functions to convert numbers to strings and strings to numbers. Some of these built-in functions have restrictions on the type of argument they take:

|Function |Description|
|-|-|
|`int`|Converts a string of digits or a float to an integer ("int") value|
|`float`|Converts a string of digits (perhaps with a decimal point) or an int to a decimal ("float") value|
|`str`|Converts any value to a string|

Try to predict what data type and value `example` evaluates to, then run the cell.

In [ ]:
example = 8 + int("10") + float("8")

print(example)
print("This example returned a " + str(type(example)) + "!")

26.0
This example returned a <class 'float'>!


Suppose you're writing a program that looks for dates in a text, and you want your program to find the amount of time that elapsed between two years it has identified.  It doesn't make sense to subtract two texts, but you can first convert the text containing the years into numbers.

**Question 1.2.1.** Finish the code below to compute the number of years that elapsed between `one_year` and `another_year`.  Don't just write the numbers `1618` and `1648` (or `30`); use a conversion function to turn the given text data into numbers.

<!--
BEGIN QUESTION
name: q121
-->

In [ ]:
# Some text data:
one_year = "1618"
another_year = "1648"

# Complete the next line.  Note that we can't just write:
#   another_year - one_year
# If you don't see why, try seeing what happens when you
# write that here.
difference = abs(int(one_year)-int(another_year))
difference

30

## 1.3. Passing strings to functions

String values, like numbers, can be arguments to functions and can be returned by functions. 

The function `len` (derived from the word "length") takes a single string as its argument and returns the number of characters (including spaces) in the string.

Note that it doesn't count *words*. `len("one small step for man")` evaluates to 22, not 5.

**Question 1.3.1.**  Use `len` to find the number of characters in the long string in the next cell.  Characters include things like spaces and punctuation. Assign `sentence_length` to that number.

(The string is the first sentence of the English translation of the French [Declaration of the Rights of Man](http://avalon.law.yale.edu/18th_century/rightsof.asp).)  

<!--
BEGIN QUESTION
name: q131
-->

In [ ]:
a_very_long_sentence = "The representatives of the French people, organized as a National Assembly, believing that the ignorance, neglect, or contempt of the rights of man are the sole cause of public calamities and of the corruption of governments, have determined to set forth in a solemn declaration the natural, unalienable, and sacred rights of man, in order that this declaration, being constantly before all the members of the Social body, shall remind them continually of their rights and duties; in order that the acts of the legislative power, as well as those of the executive power, may be compared at any moment with the objects and purposes of all political institutions and may thus be more respected, and, lastly, in order that the grievances of the citizens, based hereafter upon simple and incontestable principles, shall tend to the maintenance of the constitution and redound to the happiness of all."
sentence_length = len(a_very_long_sentence)
sentence_length

896

# 2. Arrays

Computers are most useful when you can use a small amount of code to *do the same action* to *many different things*.

For example, in the time it takes you to calculate the 18% tip on a restaurant bill, a laptop can calculate 18% tips for every restaurant bill paid by every human on Earth that day.  (That's if you're pretty fast at doing arithmetic in your head!)

**Arrays** are how we put many values in one place so that we can operate on them as a group. For example, if `billions_of_numbers` is an array of numbers, the expression

    .18 * billions_of_numbers

gives a new array of numbers that contains the result of multiplying each number in `billions_of_numbers` by .18.  Arrays are not limited to numbers; we can also put all the words in a book into an array of strings.

Concretely, an array is a **collection of values of the same type**. 

## 2.1. Making arrays

First, let's learn how to manually input values into an array. This typically isn't how programs work. Normally, we create arrays by loading them from an external source, like a data file.

To create an array by hand, call the function `np.array()`.  Each argument you pass to `np.array()` will be in the array it returns.  Run this cell to see an example of converting a list into an array:

In [ ]:
np.array([0.125, 4.75, -1.3])

Each value in an array (in the above case, the numbers 0.125, 4.75, and -1.3) is called an *element* of that array.

Arrays themselves are also values, just like numbers and strings.  That means you can assign them to names or use them as arguments to functions. For example, `len(<some_array>)` returns the number of elements in `some_array`.

**Question 2.1.1.** Make an array containing the numbers 0, 1, -1, $\pi$, and $e$, in that order.  Name it `interesting_numbers`.  

*Hint:* How did you get the values $\pi$ and $e$ in lab 2?  You can refer to them in exactly the same way here.

<!--
BEGIN QUESTION
name: q211
-->

In [ ]:
interesting_numbers = np.array([0,1,-1,math.pi,math.e])
interesting_numbers

array([ 0.        ,  1.        , -1.        ,  3.14159265,  2.71828183])

**Question 2.1.2.** Make an array containing the five strings `"Hello"`, `","`, `" "`, `"world"`, and `"!"`.  (The third one is a single space inside quotes.)  Name it `hello_world_components`.

*Note:* If you evaluate `hello_world_components`, you'll notice some extra information in addition to its contents: `dtype='<U5'`.  That's just NumPy's extremely cryptic way of saying that the data types in the array are strings.

<!--
BEGIN QUESTION
name: q212
-->

In [ ]:
hello_world_components = np.array(["Hello", ",", " ", "world", "!"])
hello_world_components

array(['Hello', ',', ' ', 'world', '!'], dtype='<U5')

###  `np.arange()`
Arrays are provided by a package called [NumPy](http://www.numpy.org/) (pronounced "NUM-pie"). The package is called `numpy`, but it's standard to rename it `np` for brevity.  You can do that with:

    import numpy as np

Very often in data science, we want to work with many numbers that are evenly spaced within some range.  NumPy provides a special function for this called `arange`.  The line of code `np.arange(start, stop, step)` evaluates to an array with all the numbers starting at `start` and counting up by `step`, stopping **before** `stop` is reached.

Run the following cells to see some examples!

In [ ]:
# This array starts at 1 and counts up by 2
# and then stops before 6
np.arange(1, 6, 2)

array([1, 3, 5])

In [ ]:
# This array doesn't contain 9
# because np.arange stops *before* the stop value is reached
np.arange(4, 9, 1)

array([4, 5, 6, 7, 8])

**Question 2.1.3.** Import `numpy` as `np` and then use `np.arange` to create an array with the multiples of 99 from 0 up to (**and including**) 9999.  (So its elements are 0, 99, 198, 297, etc.)

<!--
BEGIN QUESTION
name: q213
-->

In [ ]:
...
multiples_of_99 = np.arange(0,9999+99,99)
multiples_of_99

array([   0,   99,  198,  297,  396,  495,  594,  693,  792,  891,  990,
       1089, 1188, 1287, 1386, 1485, 1584, 1683, 1782, 1881, 1980, 2079,
       2178, 2277, 2376, 2475, 2574, 2673, 2772, 2871, 2970, 3069, 3168,
       3267, 3366, 3465, 3564, 3663, 3762, 3861, 3960, 4059, 4158, 4257,
       4356, 4455, 4554, 4653, 4752, 4851, 4950, 5049, 5148, 5247, 5346,
       5445, 5544, 5643, 5742, 5841, 5940, 6039, 6138, 6237, 6336, 6435,
       6534, 6633, 6732, 6831, 6930, 7029, 7128, 7227, 7326, 7425, 7524,
       7623, 7722, 7821, 7920, 8019, 8118, 8217, 8316, 8415, 8514, 8613,
       8712, 8811, 8910, 9009, 9108, 9207, 9306, 9405, 9504, 9603, 9702,
       9801, 9900, 9999])

##### Temperature readings
NOAA (the US National Oceanic and Atmospheric Administration) operates weather stations that measure surface temperatures at different sites around the United States.  The hourly readings are [publicly available](http://www.ncdc.noaa.gov/qclcd/QCLCD?prior=N).

Suppose we download all the hourly data from the Oakland, California site for the month of December 2015.  To analyze the data, we want to know when each reading was taken, but we find that the data don't include the timestamps of the readings (the time at which each one was taken).

However, we know the first reading was taken at the first instant of December 2015 (midnight on December 1st) and each subsequent reading was taken exactly 1 hour after the last.

**Question 2.1.4.** Create an array of the *time, in seconds, since the start of the month* at which each hourly reading was taken.  Name it `collection_times`.

*Hint 1:* There were 31 days in December, which is equivalent to ($31 \times 24$) hours or ($31 \times 24 \times 60 \times 60$) seconds.  So your array should have $31 \times 24$ elements in it.

*Hint 2:* The `len` function works on arrays, too!  If your `collection_times` isn't passing the tests, check its length and make sure it has $31 \times 24$ elements.

<!--
BEGIN QUESTION
name: q214
-->

In [ ]:
collection_times = np.arange(0,(31*21)+1,1)
collection_times

array([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,
        13,  14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,
        26,  27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,
        39,  40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,
        52,  53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,
        65,  66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,
        78,  79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,
        91,  92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103,
       104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116,
       117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129,
       130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142,
       143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155,
       156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168,
       169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 18

## 2.2. Working with single elements of arrays ("indexing")
Let's work with a more interesting dataset.  The next cell creates an array called `population_amounts` that includes estimated world populations in every year from **1950** to roughly the present.  (The estimates come from the US Census Bureau website.)

Rather than type in the data manually, we've loaded them from a file on your computer called `world_population.csv`.  You'll learn how to do that later in this lab!

In [23]:
population_amounts = pd.read_csv("world_population.csv", usecols=["Population"])
population_amounts

Population
0   2557628654
1   2594939877
2   2636772306
3   2682053389
4   2730228104
..         ...
61  6944055583
62  7022349283
63  7101027895
64  7178722893
65  7256490011

[66 rows x 1 columns]

To get the first element of `population_amounts`, we can use the `.loc()` or the `.iloc()` method, this returns the world population in the first year in the dataset, 1950.

In [ ]:
population_amounts.iloc[0]

Population    2557628654
Name: 0, dtype: int64

The value of that expression is the number 2557628654 (around 2.5 billion), because that's the first thing in the array `population_amounts`.

Notice that we wrote `.loc[0]`, not `.loc[1]`, to get the first element.  This is a weird convention in computer science.  0 is called the *index* of the first item.  It's the number of elements that appear *before* that item.  So 3 is the index of the 4th item.

Here are some more examples.  In the examples, we've given names to the things we get out of `population_amounts`.  Read and run each cell.

In [24]:
# The 13th element in the array is the population
# in 1962 (which is 1950 + 12).
population_1962 = population_amounts.loc[12]
population_1962

Population    3140093217
Name: 12, dtype: int64

In [25]:
# The 66th element is the population in 2015.
population_2015 = population_amounts.loc[65]
population_2015

Population    7256490011
Name: 65, dtype: int64

In [26]:
# The array has only 66 elements, so this doesn't work.
# (There's no element with 66 other elements before it.)
population_2016 = population_amounts.loc[66]
population_2016

KeyError: 66

Since `np.array()` returns an array, we can simply call for the element located at position `[3]` on its output to get its 4th element, just like we "chained" together calls to the method `replace` earlier.

In [ ]:
np.array([-1, -3, 4, -2])[3]

**Question 2.2.1.** Set `population_1973` to the world population in 1973, by getting the appropriate element from `population_amounts` using `item`.

<!--
BEGIN QUESTION
name: q221
-->

In [30]:
population_1973 = population_amounts.loc[13]
population_1973

Population    3209827882
Name: 13, dtype: int64

## 2.3. Doing something to every element of an array
Arrays are primarily useful for doing the same operation many times, so we don't often have to use `.item` and work with single elements.

##### Logarithms
Here is one simple question we might ask about world population:

> How big was the population in *orders of magnitude* in each year?

Orders of magnitude quantify how big a number is by representing it as the power of another number (for example, representing 104 as $10^{2.017033}$). One way to do this is by using the logarithm function. The logarithm (base 10) of a number increases by 1 every time we multiply the number by 10. It's like a measure of how many decimal digits the number has, or how big it is in orders of magnitude.

We could try to answer our question like this, using the `log10` function from the `math` module and the `item` method you just saw:

In [31]:
population_1950_magnitude = math.log10(population_amounts.loc[0])
population_1951_magnitude = math.log10(population_amounts.loc[1])
population_1952_magnitude = math.log10(population_amounts.loc[2])
population_1953_magnitude = math.log10(population_amounts.loc[3])

But this is tedious and doesn't really take advantage of the fact that we are using a computer.

Instead, NumPy provides its own version of `log10` that takes the logarithm of each element of an array.  It takes a single array of numbers as its argument.  It returns an array of the same length, where the first element of the result is the logarithm of the first element of the argument, and so on.

**Question 2.3.1.** Use `np.log10` to compute the logarithms of the world population in every year.  Give the result (an array of 66 numbers) the name `population_magnitudes`.  Your code should be very short.

<!--
BEGIN QUESTION
name: q231
-->

In [32]:
population_magnitudes = np.log10(population_amounts)
population_magnitudes

Population
0     9.407837
1     9.414127
2     9.421073
3     9.428467
4     9.436199
..         ...
61    9.841613
62    9.846482
63    9.851321
64    9.856047
65    9.860727

[66 rows x 1 columns]

What you just did is called *elementwise* application of `np.log10`, since `np.log10` operates separately on each element of the array that it's called on. Here's a picture of what's going on:

<img src="array_logarithm.jpg">


The textbook's [section](https://www.inferentialthinking.com/chapters/05/1/Arrays)  on arrays has a useful list of NumPy functions that are designed to work elementwise, like `np.log10`.

##### Arithmetic
Arithmetic also works elementwise on arrays, meaning that if you perform an arithmetic operation (like subtraction, division, etc) on an array, Python will do the operation to every element of the array individually and return an array of all of the results. For example, you can divide all the population numbers by 1 billion to get numbers in billions:

In [33]:
population_in_billions = population_amounts / 1000000000
population_in_billions

Population
0     2.557629
1     2.594940
2     2.636772
3     2.682053
4     2.730228
..         ...
61    6.944056
62    7.022349
63    7.101028
64    7.178723
65    7.256490

[66 rows x 1 columns]

You can do the same with addition, subtraction, multiplication, and exponentiation (`**`). For example, you can calculate a tip on several restaurant bills at once (in this case just 3):

In [34]:
restaurant_bills = np.array([20.12, 39.90, 31.01])
print("Restaurant bills:\t", restaurant_bills)

# Array multiplication
tips = .2 * restaurant_bills
print("Tips:\t\t\t", tips)

Restaurant bills:	 [20.12 39.9  31.01]
Tips:			 [4.024 7.98  6.202]


<img src="array_multiplication.jpg">

**Question 2.3.2.** Suppose the total charge at a restaurant is the original bill plus the tip. If the tip is 20%, that means we can multiply the original bill by 1.2 to get the total charge.  Compute the total charge for each bill in `restaurant_bills`, and assign the resulting array to `total_charges`.

<!--
BEGIN QUESTION
name: q232
-->

In [37]:
total_charges = tips+restaurant_bills
total_charges

array([24.144, 47.88 , 37.212])

**Question 2.3.3.** The array `more_restaurant_bills` contains 100,000 bills!  Compute the total charge for each one.  How is your code different?

<!--
BEGIN QUESTION
name: q233
-->

In [39]:
more_restaurant_bills = pd.read_csv("more_restaurant_bills.csv", usecols=["Bill"])
more_total_charges = (0.2*more_restaurant_bills) + more_restaurant_bills
more_total_charges

Bill
0      20.244
1      20.892
2      12.216
3      11.808
4      22.128
...       ...
99995  35.736
99996   7.440
99997  19.308
99998  18.336
99999  35.664

[100000 rows x 1 columns]

The function `sum` takes a single array of numbers as its argument.  It returns the sum of all the numbers in that array (so it returns a single number, not an array).

**Question 2.3.4.** What was the sum of all the bills in `more_restaurant_bills`, *including tips*?

<!--
BEGIN QUESTION
name: q234
-->

In [45]:
sum_of_bills = int(sum(np.array(more_restaurant_bills)))
sum_of_bills

1496441

**Question 2.3.5.** The powers of 2 ($2^0 = 1$, $2^1 = 2$, $2^2 = 4$, etc) arise frequently in computer science.  (For example, you may have noticed that storage on smartphones or USBs come in powers of 2, like 16 GB, 32 GB, or 64 GB.)  Use `np.arange` and the exponentiation operator `**` to compute the first 30 powers of 2, starting from `2^0`.

*Hint 1:* `np.arange(1, 2**30, 1)` creates an array with $2^{30}$ elements and **will crash your kernel**.

*Hint 2:* Part of your solution will involve `np.arange`, but your array shouldn't have more than 30 elements.

<!--
BEGIN QUESTION
name: q235
-->

In [53]:
powers_of_2 = (np.ones(30) *2) **np.arange(1,31,1)
powers_of_2

array([2.00000000e+00, 4.00000000e+00, 8.00000000e+00, 1.60000000e+01,
       3.20000000e+01, 6.40000000e+01, 1.28000000e+02, 2.56000000e+02,
       5.12000000e+02, 1.02400000e+03, 2.04800000e+03, 4.09600000e+03,
       8.19200000e+03, 1.63840000e+04, 3.27680000e+04, 6.55360000e+04,
       1.31072000e+05, 2.62144000e+05, 5.24288000e+05, 1.04857600e+06,
       2.09715200e+06, 4.19430400e+06, 8.38860800e+06, 1.67772160e+07,
       3.35544320e+07, 6.71088640e+07, 1.34217728e+08, 2.68435456e+08,
       5.36870912e+08, 1.07374182e+09])

## 3. Creating DataFrames (df's)

An array is useful for describing a single attribute of each element in a collection. For example, let's say our collection is all US States. Then an array could describe the land area of each state. 

Tables extend this idea by containing multiple arrays, each one describing a different attribute for every element of a collection. In this way, tables allow us to not only store data about many entities but to also contain several kinds of data about each entity.

For example, in the cell below we have two arrays. The first one, `population_amounts`, was defined above in section 2.2 and contains the world population in each year (estimated by the US Census Bureau). The second array, `years`, contains the years themselves. These elements are in order, so the year and the world population for that year have the same index in their corresponding arrays.

In [55]:
# Just run this cell

years = np.arange(1950, 2015+1)
print("Population column:", population_amounts)
print(type(population_amounts))
print()
print("Years column:", years)
print(type(years))

Population column:     Population
0   2557628654
1   2594939877
2   2636772306
3   2682053389
4   2730228104
..         ...
61  6944055583
62  7022349283
63  7101027895
64  7178722893
65  7256490011

[66 rows x 1 columns]
<class 'pandas.core.frame.DataFrame'>

Years column: [1950 1951 1952 1953 1954 1955 1956 1957 1958 1959 1960 1961 1962 1963
 1964 1965 1966 1967 1968 1969 1970 1971 1972 1973 1974 1975 1976 1977
 1978 1979 1980 1981 1982 1983 1984 1985 1986 1987 1988 1989 1990 1991
 1992 1993 1994 1995 1996 1997 1998 1999 2000 2001 2002 2003 2004 2005
 2006 2007 2008 2009 2010 2011 2012 2013 2014 2015]
<class 'numpy.ndarray'>


Suppose we want to answer this question:

> In which year did the world's population cross 6 billion?

You could technically answer this question just from staring at the arrays, but it's a bit convoluted, since you would have to count the position where the population first crossed 6 billion, then find the corresponding element in the years array. In cases like these, it might be easier to put the data into a *`Table`*, a 2-dimensional type of dataset. 



In [56]:
# Create a df 'population_amounts1' with column 'Population'
population_amounts1 = pd.read_csv("world_population.csv", usecols=["Population"])
population_amounts1.head()

Population
0  2557628654
1  2594939877
2  2636772306
3  2682053389
4  2730228104

In [57]:
# convert the numpy array 'years' to df format
years1 = pd.DataFrame(years)

# assign the outcome of rename 'years' column 0 to 'years' to a new var
years2 = years1.rename(columns={0:'years'})

In [58]:
years2.head()

years
0   1950
1   1951
2   1952
3   1953
4   1954

In [59]:
# create new df using cloumns from existing df's
population = pd.DataFrame({"Population":population_amounts1['Population'], "Years":years2['years']})
population

Population  Years
0   2557628654   1950
1   2594939877   1951
2   2636772306   1952
3   2682053389   1953
4   2730228104   1954
..         ...    ...
61  6944055583   2011
62  7022349283   2012
63  7101027895   2013
64  7178722893   2014
65  7256490011   2015

[66 rows x 2 columns]

Now the data is combined into a single table! It's much easier to parse this data. If you need to know what the population was in 1959, for example, you can tell from a single glance.

**Question 3.1.** In the cell below, we've created 2 arrays. Using the steps above, assign `top_10_movies` to a table that has two columns called "Rating" and "Name", which hold `top_10_movie_ratings` and `top_10_movie_names` respectively.

<!--
BEGIN QUESTION
name: q31
-->

In [66]:
# N.B. both data sources are numpy arrays

top_10_movie_ratings = np.array([9.2, 9.2, 9., 8.9, 8.9, 8.9, 8.9, 8.9, 8.9, 8.8])
top_10_movie_names = np.array([
        'The Shawshank Redemption (1994)',
        'The Godfather (1972)',
        'The Godfather: Part II (1974)',
        'Pulp Fiction (1994)',
        "Schindler's List (1993)",
        'The Lord of the Rings: The Return of the King (2003)',
        '12 Angry Men (1957)',
        'The Dark Knight (2008)',
        'Il buono, il brutto, il cattivo (1966)',
        'The Lord of the Rings: The Fellowship of the Ring (2001)'])

top_10_movies = pd.DataFrame({"Ratings":top_10_movie_ratings, "Names":top_10_movie_names})

# We've put this next line here 
# so your table will get printed out 
# when you run this cell.
top_10_movies

Ratings                                              Names
0      9.2                    The Shawshank Redemption (1994)
1      9.2                               The Godfather (1972)
2      9.0                      The Godfather: Part II (1974)
3      8.9                                Pulp Fiction (1994)
4      8.9                            Schindler's List (1993)
5      8.9  The Lord of the Rings: The Return of the King ...
6      8.9                                12 Angry Men (1957)
7      8.9                             The Dark Knight (2008)
8      8.9             Il buono, il brutto, il cattivo (1966)
9      8.8  The Lord of the Rings: The Fellowship of the R...

#### Loading a table from a file

In most cases, we aren't going to go through the trouble of typing in all the data manually. Instead, we load them in from an external source, like a data file. There are many formats for data files, but CSV ("comma-separated values") is the most common.

`pd.read_csv(...)` takes one argument (a path to a data file in string format) and returns a table.  

**Question 3.2.** `imdb.csv` contains a table of information about the 250 highest-rated movies on IMDb.  Load it as a table called `imdb`.

(You may remember working with this table in Lab 2!)


In [64]:
imdb = pd.read_csv("imdb.csv")
imdb

Votes  Rating                                  Title  Year  Decade
0      88355     8.4                                      M  1931    1930
1     132823     8.3                    Singin' in the Rain  1952    1950
2      74178     8.3                          All About Eve  1950    1950
3     635139     8.6                                   Léon  1994    1990
4     145514     8.2                       The Elephant Man  1980    1980
..       ...     ...                                    ...   ...     ...
245  1078416     8.7                           Forrest Gump  1994    1990
246    31003     8.1                  Le salaire de la peur  1953    1950
247   167076     8.2                               3 Idiots  2009    2000
248    91689     8.1                                Network  1976    1970
249   589477     8.3  Eternal Sunshine of the Spotless Mind  2004    2000

[250 rows x 5 columns]

Where did `imdb.csv` come from? Take a look at [this lab's folder](./). You should see a file called `imdb.csv`.

Open up the `imdb.csv` file in that folder and look at the format. What do you notice? The `.csv` filename ending says that this file is in the [CSV (comma-separated value) format](http://edoceo.com/utilitas/csv-file-format).

## 4. More Table Operations!

Now that you've worked with arrays, let's add a few more methods to the list of table operations that you saw in Lab 2.

### `df['column']`

`df['column']` takes the column name of a table (in string format) as its argument and returns the values in that column as an **array**. 

In [67]:
# Returns an array of movie names
top_10_movies['Names']

0                      The Shawshank Redemption (1994)
1                                 The Godfather (1972)
2                        The Godfather: Part II (1974)
3                                  Pulp Fiction (1994)
4                              Schindler's List (1993)
5    The Lord of the Rings: The Return of the King ...
6                                  12 Angry Men (1957)
7                               The Dark Knight (2008)
8               Il buono, il brutto, il cattivo (1966)
9    The Lord of the Rings: The Fellowship of the R...
Name: Names, dtype: object

### `take`
The table method `take` takes as its argument as an array of numbers.  Each number should be the index of a row in the table.  It returns a **new table** with only those rows. 

You'll usually want to use `take` in conjunction with `np.arange` to take the first few rows of a table.

In [69]:
# Take first 5 movies of top_10_movies
top_10_movies.take(np.arange(0, 5, 1))

Ratings                            Names
0      9.2  The Shawshank Redemption (1994)
1      9.2             The Godfather (1972)
2      9.0    The Godfather: Part II (1974)
3      8.9              Pulp Fiction (1994)
4      8.9          Schindler's List (1993)

The next three questions will give you practice with combining the operations you've learned in this lab and the previous one to answer questions about the `population` and `imdb` tables. First, check out the `population` table from section 2.

In [68]:
# Run this cell to display the population table.
population

Population  Years
0   2557628654   1950
1   2594939877   1951
2   2636772306   1952
3   2682053389   1953
4   2730228104   1954
..         ...    ...
61  6944055583   2011
62  7022349283   2012
63  7101027895   2013
64  7178722893   2014
65  7256490011   2015

[66 rows x 2 columns]

**Question 4.1.** Check out the `population` table from section 2 of this lab.  Compute the year when the world population first went above 6 billion. Assign the year to `year_population_crossed_6_billion`.

<!--
BEGIN QUESTION
name: q41
-->

In [72]:
year_population_crossed_6_billion = sorted(population["Years"][population["Population"]>6000000000])[0]
year_population_crossed_6_billion

1999

**Question 4.2.** Find the average rating for movies released before the year 2000 and the average rating for movies released in the year 2000 or after for the movies in `imdb`.

*Hint*: Think of the steps you need to do (take the average, find the ratings, find movies released in 20th/21st centuries), and try to put them in an order that makes sense.

<!--
BEGIN QUESTION
name: q42
-->

In [73]:
before_2000 = sum(imdb["Rating"][imdb["Year"]<2000])/len(imdb["Rating"][imdb["Year"]<2000])
after_or_in_2000 = sum(imdb["Rating"][imdb["Year"]>=2000])/len(imdb["Rating"][imdb["Year"]>=2000])
print("Average before 2000 rating:", before_2000)
print("Average after or in 2000 rating:", after_or_in_2000)

Average before 2000 rating: 8.278362573099418
Average after or in 2000 rating: 8.237974683544303


**Question 4.3.** Here's a challenge: Find the number of movies that came out in *even* years.

*Hint:* The operator `%` computes the remainder when dividing by a number.  So `5 % 2` is 1 and `6 % 2` is 0.  A number is even if the remainder is 0 when you divide by 2.

*Hint 2:* `%` can be used on arrays, operating elementwise like `+` or `*`.  So `make_array(5, 6, 7) % 2` is `array([1, 0, 1])`.

*Hint 3:* Create a column called "Year Remainder" that's the remainder when each movie's release year is divided by 2.  Make a copy of `imdb` that includes that column (`imdb.with_column(...)` returns a new table).  Then use `where` to find rows where that new column is equal to 0.  Then use `num_rows` to count the number of such rows.

*Note:* These steps can be chained in one single statement, or broken up across several lines with intermediate names assigned. You’re always welcome to break down problems however you wish!

<!--
BEGIN QUESTION
name: q43
-->

In [76]:
num_even_year_movies = imdb[(imdb["Year"]%2)==0]
num_even_year_movies

Votes  Rating                                  Title  Year  Decade
1     132823     8.3                    Singin' in the Rain  1952    1950
2      74178     8.3                          All About Eve  1950    1950
3     635139     8.6                                   Léon  1994    1990
4     145514     8.2                       The Elephant Man  1980    1980
6     441174     8.1                              Gone Girl  2014    2010
..       ...     ...                                    ...   ...     ...
240   117590     8.4                     The Great Dictator  1940    1940
242   476501     8.6                         Cidade de Deus  2002    2000
245  1078416     8.7                           Forrest Gump  1994    1990
248    91689     8.1                                Network  1976    1970
249   589477     8.3  Eternal Sunshine of the Spotless Mind  2004    2000

[127 rows x 5 columns]

Congratulations, you're done with lab 3!  Be sure to 
- **run all the tests** (the next cell has a shortcut for that), 
- **Save and Checkpoint** from the `File` menu,
- **run the last cell to submit your work**,
- and ask one of the staff members to check you off.